### PR TITLE
[PDI-19665] Hadoop Copy Files exposes password from Hadoop Cluster in…

### PR DESCRIPTION
**Pull request by NJtwentyone**
… clear-text on destination URL

- allowing cleartext passwords in all server based URLs in JobEntryCopyFiles related job entries
- adding function saveURL() as companion to loadURL()

**background on re-opening issue:** https://hv-eng.atlassian.net/browse/PDI-19665?focusedCommentId=2033565

**technical background:** https://hv-eng.atlassian.net/browse/PDI-19665?focusedCommentId=2033576

**merge order dependency**

1. https://github.com/pentaho/pentaho-kettle/pull/9469
2. https://github.com/pentaho/big-data-plugin/pull/2560

### CODE REVIEW

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 25m 55s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl engine
```
---
:ok_hand: All tests passed!
 
Tests run: **2325**, Failures: **0**, Skipped: **13** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9469/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by NJtwentyone** [comment]
There are two code smells which is persisted not to to address

1. unused parameters in `saveURL( String url, String ncName, IMetaStore metastore, Map<String, String> mappings ) {`. I purposely did that to match loadURL and for future proofing the function.
2. _literal "wildcard"_ . All this code is legacy. I already addressed one sonarqube findings by creating a variable XML_INDENTATION. the real issue is we are hand writing XML and addressing that is out of scope of the PR.

File: engine/src/main/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFiles.java [hunk 1]
```diff
@@ -93,6 +93,8 @@ public class JobEntryCopyFiles extends JobEntryBase implements Cloneable, JobEnt
   public static final String DEST_URL = "EMPTY_DEST_URL-";
   public static final String SOURCE_URL = "EMPTY_SOURCE_URL-";
 
+  private static final String XML_INDENTATION = "          ";
```
Comment (NJtwentyone) [inline engine/src/main/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFiles.java:96]: added to get rid of critical sonar qube warning. The real issue this legacy code is hand writting xml, and addressing that is out of scope of this PR.

File: engine/src/main/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFiles.java [hunk 2]
```diff
@@ -232,15 +234,16 @@ protected String loadDestination( Node fnode ) {
...
+    String saveURL = saveURL( source, namedCluster, getMetaStore(), configurationMappings  ) ;
+    retval.append( XML_INDENTATION ).append( XMLHandler.addTagValue( SOURCE_FILE_FOLDER, saveURL ) );
+    retval.append( XML_INDENTATION ).append( XMLHandler.addTagValue( SOURCE_CONFIGURATION_NAME, namedCluster ) );
   }
 
   protected void saveDestination( StringBuilder retval, String destination ) {
     String namedCluster = configurationMappings.get( destination );
-    retval.append( "          " ).append( XMLHandler.addTagValue( DESTINATION_FILE_FOLDER, KettleVFS.cleanseFilename( destination ) ) );
-    retval.append( "          " ).append( XMLHandler.addTagValue( DESTINATION_CONFIGURATION_NAME, namedCluster ) );
+    String saveURL = saveURL( destination, namedCluster, getMetaStore(), configurationMappings  ) ;
```
Comment (NJtwentyone) [inline engine/src/main/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFiles.java:244]: The Hadoop Copy files job entry code will have logic to restore the original URL, which doesn't have the user credentials information in the URL.

File: engine/src/main/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFiles.java [hunk 3]
```diff
@@ -1155,13 +1160,34 @@ public boolean evaluates() {
...
 
+  /**
+   * Marshalling the actual URL to job entry xml fields ( {@link #SOURCE_FILE_FOLDER } and {@link #DESTINATION_FILE_FOLDER} )
+   * @param url
+   * @param ncName
+   * @param metastore
+   * @param mappings
+   * @return
+   */
+  public String saveURL( String url, String ncName, IMetaStore metastore, Map<String, String> mappings ) {
```
Comment (NJtwentyone) [inline engine/src/main/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFiles.java:1187]: see other PR for why is is necessary https://github.com/pentaho/big-data-plugin/pull/2560

File: engine/src/test/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFilesTest.java [hunk 4]
```diff
@@ -151,22 +153,20 @@ public void saveLoad() throws Exception {
 
   @Test
   public void saveLoadWithPassword() throws Exception {
-    String[] srcPath = new String[] { "EMPTY_SOURCE_URL-0-hdfs://user321:321fake@foo.bar.com:8020/user/user321" };
-    String[] destPath = new String[] { "EMPTY_DEST_URL-0-hdfs://user123:fake123@foo.bar.com:8020/user/user123" };
-
-    String srcPathScrubbedEncoded = "EMPTY_SOURCE_URL-0-hdfs://user321:&lt;password&gt;@foo.bar.com:8020/user/user321";
-    String destPathScrubbedEncoded = "EMPTY_DEST_URL-0-hdfs://user123:&lt;password&gt;@foo.bar.com:8020/user/user123";
+    String srcPath = "EMPTY_SOURCE_URL-0-hdfs://user321:321fake@foo.bar.com:8020/user/user321";
```
Comment (NJtwentyone) [inline engine/src/test/java/org/pentaho/di/job/entries/copyfiles/JobEntryCopyFilesTest.java:156]: purposely allowing hardcoded URLs, this test verifies this, see "Technical Direction" section of jira comment : https://hv-eng.atlassian.net/browse/PDI-19665?focusedCommentId=2033576

**Review by peterrinehart — APPROVED** [review]

**Review by tkafalas — APPROVED** [review]

**Review by rmansoor — APPROVED** [review]
